src/EFCore.Cosmos/Metadata/Conventions/CosmosJsonIdConvention.cs

```diff
@@ -0,0 +1,270 @@
+            {
+                if (jsonIdProperty != null
+                    && !jsonIdProperty.GetConfigurationSource().OverridesStrictly(ConfigurationSource.Convention))
+                {
+                    jsonIdProperty.Builder.ToJsonProperty(null);
+                }
+
+                if (computedIdProperty != null
+                    && !computedIdProperty.GetConfigurationSource().OverridesStrictly(ConfigurationSource.Convention))
+                {
+                    entityTypeBuilder.Metadata.RemoveProperty(computedIdProperty);
+                }
+
+                return;
+            }
+
+            // Reactions required for:
+            // IEntityTypeAnnotationChangedConvention (AlwaysCreateShadowIdProperty)
+            // IModelAnnotationChangedConvention (AlwaysCreateShadowIdProperty)
+            var alwaysCreateId = entityType.GetAlwaysCreateShadowIdProperty() ?? entityType.Model.GetAlwaysCreateShadowIdProperty();
+            if (alwaysCreateId != true)
+            {
+                // If there is one string primary key property after removing partition keys, then map it to the JSON id field directly,
+                // unless it is explicitly mapped to some other property, in which case we compute the field value as below.
+
+                // IKeyAddedConvention, IKeyRemovedConvention, IPropertyAddedConvention, IPropertyRemovedConvention,
+                // IEntityTypeAnnotationChangedConvention (PartitionKeyNames) (DiscriminatorInKey)
+                // IDiscriminatorPropertySetConvention
+                var idDefinition = DefinitionFactory.Create((IEntityType)entityType)!;
+                var keyProperty = (IConventionProperty?)idDefinition.Properties.FirstOrDefault();
+                if (idDefinition.DiscriminatorEntityType == null
+                    && idDefinition.Properties.Count == 1)
+                {
+                    var clrType = keyProperty!.GetValueConverter()?.ProviderClrType ?? keyProperty.ClrType;
+                    if (clrType == typeof(string))
+                    {
+                        if (computedIdProperty != null)
+                        {
+                            entityTypeBuilder.Metadata.RemoveProperty(computedIdProperty);
+                        }
+                        keyProperty.SetJsonPropertyName(IdPropertyJsonName);
+                        return;
+                    }
+                }
+            }
+
+            if (jsonIdProperty != null
+                && jsonIdProperty != computedIdProperty)
+            {
+                jsonIdProperty.Builder.ToJsonProperty(null);
+            }
+
+            computedIdProperty = entityTypeBuilder
+                .Property(typeof(string), DefaultIdPropertyName, setTypeConfigurationSource: false)!
+                .ToJsonProperty(IdPropertyJsonName)!
+                .IsRequired(true)!
+                .HasValueGeneratorFactory(typeof(IdValueGeneratorFactory))!
+                .Metadata;
```

Comment:
All of these can return `null`
```suggestion
            computedIdProperty = entityTypeBuilder
                .Property(typeof(string), DefaultIdPropertyName, setTypeConfigurationSource: false)
                ?.ToJsonProperty(IdPropertyJsonName)
                ?.IsRequired(true)
                ?.HasValueGeneratorFactory(typeof(IdValueGeneratorFactory))
                ?.Metadata;
```